cyten/planar.py

```diff
@@ -0,0 +1,937 @@
+"""TODO"""
+# Copyright (C) TeNPy Developers, Apache license
+from __future__ import annotations
+from typing import Literal, Hashable, Sequence
+import warnings
+
+import numpy as np
+
+from .tensors import (
+    Tensor, LabelledLegs, partial_trace, tdot, is_valid_leg_label,
+    CONTRACT_SYMBOL, LEG_SELECT_SYMBOL, OPEN_LEG_SYMBOL
+)
+from .sparse import LinearOperator
+from .tools import duplicate_entries, BigOPolynomial, to_iterable, to_valid_idx
+
+
+type NestedContainer[T] = tuple[T | NestedContainer[T], ...] | list[T | NestedContainer[T]]
+"""Type variable for nested containers (lists, tuples or mixed) that contain type ``T``"""
+
+
+class TensorPlaceholder(LabelledLegs):
+    """Placeholder for a tensor used to define :class:`PlanarDiagram` s.
+
+    Attributes
+    ----------
+    labels : list of str
+        The labels of the tensor (up to cyclic permutation). This means that as long as we go
+        clockwise around the shape, any starting point can be chosen for the labels.
+    dims : list of (str | None)
+        For each of the legs, an optional symbol to represent its dimension.
+    """
+
+    def __init__(self, labels: list[str], dims: list[str] = None,
+                 cost_to_make: BigOPolynomial = BigOPolynomial()):
```

Comment:
the default argument for cost_to_make is currently shared between different instances...